### PR TITLE
Fix shutdown race condition

### PR DESCRIPTION
**Pull request by righdforsa**
Note: rebased version of https://github.com/Expensify/Bedrock/pull/1137

### Details
To prevent a failed shutdown when Auth servers switch from STANDINGDOWN to SEARCHING with commands in the early part of the "processCommand" function, we need to catch an additional edge case, otherwise we get stuck with the command endlessly retrying.

### Fixed Issues
Fixes [Expensify/Expensify#176212](https://github.com/Expensify/Expensify/issues/176212)

### Tests
Ran GracefulFailoverTest repeatedly.

To reproduce the failure, I changed the Graceful shutdown timeout _gracefulShutdownTimeout.alarmDuration from 60s to 15s and added a random 30s sleep to the top of BedrockCore::processCommand. This fix resolved that issue. Ran the test 10x to make sure it was good and fixed.
Unwound the 15s shutdown timeout and ran the test again 10x to make sure it was still good.
Unwound the 30s random processCommand timeout and all debug logging and ran the test again 10x to make sure it was still good. Interestingly, this one had a shutdown failure in one of the follower nodes on the 7th test pass, but it was in a different part of the code, so I believe it's distinct, and a repeat of the tests passed all 10x.

Had issues with BadCommandTest and StatusHandlingCommandsTest failing intermittently. Most recent test I did 10x with "-only BadCommand,StatusHandlingCommands" and "-repeatCount 10", so 100 total runs, with no issues. Testing with full clustertest again now.
_________
**Internal Testing Reminder:** when changing bedrock, please compile auth against your new changes

### CODE REVIEW

**Comment by righdforsa** [comment]
Since BadCommandTest failed twice, and StatusHandlingCommands failed in Travis, (and locally with some prodding) and both of them failed in ASSERT statements regarding SQLiteNode states (e.g. LEADING, FOLLOWING, STANDINGDOWN) I'm going to take these as "things that need to be identified or fixed to move forward". Working on them.

File: BedrockServer.cpp [hunk 1]
```diff
@@ -1079,20 +1079,28 @@ void BedrockServer::worker(int threadId)
...
+                                SINFO("State changed before 'processCommand' and we're shutting down. Abandoning " << command->request.methodLine << ".");
+                                command->response.methodLine = "500 Leader stopped leading";
+                                command->complete = true;
+                            // If we're not shutting down, we can likely recover from the current state changes.
                             // Allow for an extra retry and start from the top, like with ABANDONED_FOR_CHECKPOINT.
-                            SINFO("State changed before 'processCommand' but no HTTPS requests so retrying.");
-                            ++retry;
+                            } else {
+                                SINFO("State changed before 'processCommand' but no HTTPS requests so retrying.");
+                                ++retry;
```
Comment (flodnv) [inline BedrockServer.cpp:1102]: Discussed about this with @righdforsa:

Why would we keep this `++retry;` here -- in which situation would retrying infinitely (see new line 1125) would ever benefit us? This log line hasn't been logged in over 1 week. Can we not just give this very rare flow the regular amount of retries @tylerkaraszewski ? `++retry` in `ABANDONED_FOR_CHECKPOINT` conceptually makes theoretical sense, but I just can't make any sense of it here.

Comment (tylerkaraszewski) [reply BedrockServer.cpp:1102]: We did not count any of the exception cases (like this one, the state changed while we were running the command) toward the retry count. The purpose of the retry count is to avoid conflicts, when we run out of retries, we push the command onto the blocking commit queue. Why should this state change affect the conflict-handling behavior?

Yes, we could remove it and it would likely make little difference but we could also leave it and it would likely make little difference because it's a weird exception handling case that almost never happens.

Sending the command to the blocking commit queue is actually really weird if we have just changed from LEADING to FOLLOWING (I think it will actually still work, though).

Comment (flodnv) [reply BedrockServer.cpp:1102]: >we could also leave it and it would likely make little difference because it's a weird exception handling case that almost never happens.

Right, but my point is that this can result, and has resulted, in infinite loops. So if we can just avoid this possibility, we're not really making anything worse, are we?

Comment (tylerkaraszewski) [reply BedrockServer.cpp:1102]: Are we sure this in particular has resulted in an infinite loop? I'm not actually sure how that would happen.

Comment (flodnv) [reply BedrockServer.cpp:1102]: @righdforsa seemed sure when he explained it to me yesterday 😅

Comment (tylerkaraszewski) [reply BedrockServer.cpp:1102]: I did not hear that explanation. :)

Comment (righdforsa) [reply BedrockServer.cpp:1102]: > Are we sure this in particular has resulted in an infinite loop? I'm not actually sure how that would happen.

That's the whole point of the ticket that spawned this PR. When we got into shutdown, there was a command running that happened to take just long enough that it began processing before the leader node got past standingdown but when "processCommand()" hit the "is this node leading?" block, it was no longer leading, and returned a "not leading" result. Because of the not leading result, we hit this retry increment, and the command looped over and over and never recovered, which caused the sync thread to not be able to shut down, which left the node port open, which caused peers to think it was still available, and caused a 5 minute site outage.

Comment (righdforsa) [reply BedrockServer.cpp:1102]: Now that that condition is handled (in this PR) in order to get there again, we'd have to disconnect from the cluster but not be shutting down. I don't know if we want to loop endlessly in that case.

Comment (tylerkaraszewski) [reply BedrockServer.cpp:1102]: > Because of the not leading result, we hit this retry increment, and the command looped over and over and never recovered

Ah, looking at it now, I see how it could do that. That's not how it's intended to work. The reason it gets stuck in the loop is because it doesn't notice it's not leading at the top of the loop.

This block of code determines if this command can write:
https://github.com/Expensify/Bedrock/blob/master/BedrockServer.cpp#L924-L946

But it's only run before the retry loop.

We could move it inside the loop, or we could just set `canWriteParallel` to false when we notice we're no longer leading.

**Comment by flodnv** [comment]
bump @righdforsa ?

**Comment by righdforsa** [comment]
I'm actively working on getting this up to date with the current state of
the Bedrock repo. Unfortunately, I've been having some stability problems
with the clustertest tests that I've been trying to resolve. This fix
works, but it's not totally clear to me what is being proposed as the
alternative approach, so maybe it's better.
